File: src/main/java/hacker/solution/OneDArrayGame.java
package hacker.solution;

import java.util.Scanner;

public class OneDArrayGame {
    private static boolean isSolvable(int [] array, int m, int i) {
        /* Base Cases */
        if (i < 0 || array[i] == 1) {
            return false;
        } else if (i + 1 >= array.length || i + m >= array.length) {
            return true;
        }
        array[i] = 1; // marks as visited
        /* Recursive Cases */
        return isSolvable(array, m, i + m) ||
                isSolvable(array, m, i + 1) ||
                isSolvable(array, m, i - 1);
    }
    private static boolean solvable(int[] array, int leap, int j) {
        int[] used = new int[array.length];
        int i=0;
        boolean flag = true;
        while (flag) {
            if (used[i]==1) return false;
            used[i]=1;
            if (i<0 || array[i]==1) return false;
            if (i+1 >= array.length || i+leap >= array.length) return true;
            if (array[i+leap]==0) i=i+leap;
            else if (array[i+1]==0) i=i+1;
            else if (array[i-1]==0) i=i-1;
        }
        return flag;
    }
    public static boolean canWin(int leap, int[] game) {
        // Return true if you can win the game; otherwise, return false.
        return solvable(game,leap,0);
    }
    public static void main(String[] args) {
        Scanner scan = new Scanner(System.in);
        int q = scan.nextInt();
        while (q-- > 0) {
            int n = scan.nextInt();
            int leap = scan.nextInt();

            int[] game = new int[n];
            for (int i = 0; i < n; i++) {
                game[i] = scan.nextInt();
            }

            System.out.println( (canWin(leap, game)) ? "YES" : "NO" );
        }
        scan.close();
    }
}

File: src/main/kotlin/hacker/solutionk/OneDArrayGame.kt
package hacker.solutionk

import java.util.*

object OneDArrayGame {
    private fun canWin(leap: Int, game: IntArray): Boolean {
        var used = IntArray(game.size)
        for (i in 0 until game.size) used[i]=0
        var position=0;
        var flag: Boolean = true;
        while(flag) {
            if (used[position] == 1) return false;
            used[position]=1;
            if (position < 0 || game[position] == 1) return false
            if (position+leap >= game.size || position+1 >= game.size) return true;
            if (game[position+leap] == 0) position += leap
            else if (game[position+1] == 0) position += 1
            else if (game[position-1] == 0) position -= 1
        }
        return flag
    }
    @JvmStatic
    fun main(args: Array<String>) {
        val scan = Scanner(System.`in`)
        var q = scan.nextInt()
        while (q-- > 0) {
            val n = scan.nextInt()
            val leap = scan.nextInt()

            val game = IntArray(n)
            for (i in 0 until n) {
                game[i] = scan.nextInt()
            }

            println(if (canWin(leap, game)) "YES" else "NO")
        }
        scan.close()
    }
}